audioutils/alsa-lib/pcm.c

```diff
@@ -0,0 +1,808 @@
+{
+  return 1;
+}
+
+int snd_pcm_poll_descriptors(snd_pcm_t *pcm, struct pollfd *pfds,
+                             unsigned int space)
+{
+  assert(pcm && pfds);
+  return pcm->ops->poll_descriptors(pcm->ops_arg, pfds, space);
+}
+
+snd_pcm_sframes_t snd_pcm_writei(snd_pcm_t *pcm, const void *buffer,
+                                 snd_pcm_uframes_t size)
+{
+  CHECK_PCM_SETUP(pcm);
+
+  return pcm->ops->writei(pcm->ops_arg, buffer, size);
+}
+
+int snd_pcm_set_volume(FAR snd_pcm_t *pcm, int volume)
+{
+  if (volume < 0 || volume > 100)
+    {
+      return -EINVAL;
+    }
+
+  pcm->volume = volume / 100.0;
+  SND_INFO("set volume:%f", pcm->volume);
+
+  return 0;
+}
+
+int snd_pcm_set_volume_db(FAR snd_pcm_t *pcm, int db)
+{
+  if (db > 0)
+    {
+      return -EINVAL;
+    }
+
+  pcm->volume = powf(10.0,  (double)db / 2000.0);
+  SND_INFO("set volume:%f", pcm->volume);
+
+  return 0;
+}
+
+int snd_pcm_set_vis_fifo(FAR snd_pcm_t *pcm, FAR const char *fifo_name)
+{
+  if (!fifo_name)
+    {
+      return -EINVAL;
+    }
+
+  if (mkfifo(fifo_name, 0666) < 0 && errno != EEXIST)
+    {
+      SND_INFO("mkfifo error: %s\n", strerror(errno));
+      return -errno;
+    }
+
+  pcm->vis_fd = open(fifo_name, O_RDONLY | O_NONBLOCK, 0);
+  if (pcm->vis_fd < 0)
+    {
+      SND_INFO("open fifo error: %s\n", strerror(errno));
+      return -errno;
+    }
+
+  SND_INFO("open vis fifo:%s, %d", fifo_name, pcm->vis_fd);
+
+  return 0;
+}
+
+int snd_pcm_close_vis_fifo(FAR snd_pcm_t *pcm)
+{
+  if (pcm->vis_fd >= 0)
+    {
+      close(pcm->vis_fd);
+      pcm->vis_fd = -1;
+    }
+
+  return 0;
+}
+
+FAR const char *snd_pcm_stream_name(snd_pcm_stream_t stream)
+{
+  if (stream > SND_PCM_STREAM_LAST)
+    {
+      return NULL;
+    }
+
+  const char *_snd_pcm_stream_names[] =
+    {
+      STREAM(PLAYBACK),
+      STREAM(CAPTURE),
+    };
+
+  return _snd_pcm_stream_names[stream];
+}
+
+FAR const char *snd_pcm_access_name(snd_pcm_access_t acc)
+{
+  if (acc > SND_PCM_ACCESS_LAST)
+    {
+      return NULL;
+    }
+
+  const char *_snd_pcm_access_names[] =
+    {
+      ACCESS(MMAP_INTERLEAVED),  ACCESS(MMAP_NONINTERLEAVED),
+      ACCESS(MMAP_COMPLEX),      ACCESS(RW_INTERLEAVED),
+      ACCESS(RW_NONINTERLEAVED),
+    };
+
+  return _snd_pcm_access_names[acc];
+}
+
+FAR const char *snd_pcm_format_name(snd_pcm_format_t format)
+{
+  if (format > SND_PCM_FORMAT_LAST)
+    {
+      return NULL;
+    }
+
+  const char *_snd_pcm_format_names[] =
+    {
+      FORMAT(S8),     FORMAT(U8),     FORMAT(S16_LE), FORMAT(S16_BE),
+      FORMAT(U16_LE), FORMAT(U16_BE), FORMAT(S24_LE), FORMAT(S24_BE),
+      FORMAT(U24_LE), FORMAT(U24_BE), FORMAT(S32_LE), FORMAT(S32_BE),
+      FORMAT(U32_LE), FORMAT(U32_BE),
+    };
+
+  return _snd_pcm_format_names[format];
+}
+
+FAR const char *snd_pcm_format_description(snd_pcm_format_t format)
+{
+  if (format > SND_PCM_FORMAT_LAST)
+    {
+      return NULL;
+    }
+
+  const char *_snd_pcm_format_descriptions[] =
```

Comment:
this file looks like highly inspired by https://github.com/alsa-project/alsa-lib/blob/master/src/pcm/pcm.c#L2001

Reply:
in order to maintain compatibility?

Reply:
I doubt that compatibility is a sufficient reason not to name this implementation as "derivative GPL work".
Even internal static functions are called the same like in alsa-lib.